npm/pkg/controlplane/controllers/v2/podcontroller.go

```diff
@@ -0,0 +1,633 @@
+		if err = c.dp.RemoveFromSet([]string{podIPSetName}, cachedNpmPod.PodIP, podKey); err != nil {
+			return fmt.Errorf("[syncAddAndUpdatePod] Error: failed to delete pod from label ipset with err: %v", err)
+		}
+		// {IMPORTANT} The order of compared list will be key and then key+val. NPM should only append after both key
+		// key + val ipsets are worked on. 0th index will be key and 1st index will be value of the label
+		removedLabelKey, removedLabelValue := util.GetLabelKVFromSet(podIPSetName)
+		if removedLabelValue != "" {
+			cachedNpmPod.removeLabelsWithKey(removedLabelKey)
+		}
+	}
+
+	// Add the pod to its label's ipset.
+	for _, addIPSetName := range addToIPSets {
+		klog.Infof("Adding pod %s to ipset %s", newPodObj.Status.PodIP, addIPSetName)
+		if err = c.dp.AddToSet([]string{addIPSetName}, newPodObj.Status.PodIP, podKey); err != nil {
+			return fmt.Errorf("[syncAddAndUpdatePod] Error: failed to add pod to label ipset with err: %v", err)
+		}
+		// {IMPORTANT} Same as above order is assumed to be key and then key+val. NPM should only append to existing labels
+		// only after both ipsets for a given label's key value pair are added successfully
+		// (TODO) will need to remove this ordering dependency
+		addedLabelKey, addedLabelValue := util.GetLabelKVFromSet(addIPSetName)
+		if addedLabelValue != "" {
+			cachedNpmPod.appendLabels(map[string]string{addedLabelKey: addedLabelValue}, AppendToExistingLabels)
+		}
+	}
+	// This will ensure after all labels are worked on to overwrite. This way will reduce any bugs introduced above
+	// If due to ordering issue the above deleted and added labels are not correct,
+	// this below appendLabels will help ensure correct state in cache for all successful ops.
+	cachedNpmPod.appendLabels(newPodObj.Labels, ClearExistingLabels)
+
+	// (TODO): optimize named port addition and deletions.
+	// named ports are mostly static once configured in todays usage pattern
+	// so keeping this simple by deleting all and re-adding
+	newPodPorts := getContainerPortList(newPodObj)
+	if !reflect.DeepEqual(cachedNpmPod.ContainerPorts, newPodPorts) {
+		// Delete cached pod's named ports from its ipset.
+		if err = c.manageNamedPortIpsets(
+			cachedNpmPod.ContainerPorts, podKey, cachedNpmPod.PodIP, deleteNamedPort); err != nil {
+			return fmt.Errorf("[syncAddAndUpdatePod] Error: failed to delete pod from named port ipset with err: %v", err)
+		}
+		// Since portList ipset deletion is successful, NPM can remove cachedContainerPorts
+		cachedNpmPod.removeContainerPorts()
+
+		// Add new pod's named ports from its ipset.
+		if err = c.manageNamedPortIpsets(newPodPorts, podKey, newPodObj.Status.PodIP, addNamedPort); err != nil {
+			return fmt.Errorf("[syncAddAndUpdatePod] Error: failed to add pod to named port ipset with err: %v", err)
+		}
+		cachedNpmPod.appendContainerPorts(newPodObj)
+	}
+	cachedNpmPod.updateNpmPodAttributes(newPodObj)
+
```

Comment:
Along with ApplyDataplane, we should be calling "UpdatePod" for both Add and Update pod, delete we do not care as the rules will be removed